File: clean_code_study/230222_01_fp_oop_dependency_inversion/java/dip_way_oop/Main.java

class Main {
    public static void main(String[] args) {
        // calling context
        HttpClient client = new HttpClient();
        App.start(client);
    }
}

class App {
    public static void start(ClientInterface client){
        client.getUser("bob@clean.com");
        User user = new User("Bob", "bob@clean.com");
        client.createUser(user);
    }
}

interface ClientInterface {
    public void getUser(String email);
    public void createUser(User user);
}

class HttpClient implements ClientInterface {

    public HttpClient(){
    }

    public void getUser(String email){
        System.out.println("Getting user... " + email);
    };

    public void createUser(User user){
        System.out.println("Creating user... " + user.name);
    };
}

class User{
    public String name;
    public String email;

    public User(String name, String email){
        this.name = name;
        this.email = email;
    }
}

File: clean_code_study/230222_01_fp_oop_dependency_inversion/go/dip_way_oop/main.go
package main

import "fmt"

func main(){
	// calling phase
	client := HttpClient{}
	app(client)
}

// declaration phase
func app(client ClientInterface){
	client.getUser("bob@clean.com")
	user := User{
		name: "Bob",
		email: "bob@clean.com",
	} 
	client.createUser(user)
}

type ClientInterface interface {
	getUser(email string) 
	createUser(user User)
}

type HttpClient struct {
}

func (c HttpClient) getUser(email string){
	fmt.Println("Getting user...", email)
}

func (c HttpClient) createUser(user User){
	fmt.Println("Creating user...", user.name)
}

type User struct {
	name string
	email string
}